tlv/stream_test.go

```diff
@@ -0,0 +1,408 @@
+package tlv_test
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"reflect"
+	"testing"
+
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/tlv"
+	"github.com/lightningnetwork/lnd/watchtower/blob"
+	"github.com/lightningnetwork/lnd/watchtower/wtwire"
+)
+
+type testObj struct {
+	u16 uint16
+	u32 uint32
+	u64 uint64
+	buf []byte
+}
+
+type tlvTestCase struct {
+	name   string
+	encGen func() (*testObj, *tlv.Stream)
+	decGen func() (*testObj, *tlv.Stream)
+	expObj *testObj
+}
+
+var tlvTests = []tlvTestCase{
+	{
+		name: "empty",
+		encGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+			)
+			return o, s
+		},
+		decGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(2, &o.u32),
+				tlv.MakePrimitiveRecord(3, &o.u64),
+				tlv.MakePrimitiveRecord(4, &o.buf),
+			)
+			return o, s
+		},
+		expObj: &testObj{},
+	},
+	{
+		name: "partial encode",
+		encGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{
+				u32: 3,
+				buf: bytes.Repeat([]byte{0x05}, 3),
+			}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(2, &o.u32),
+				tlv.MakePrimitiveRecord(4, &o.buf),
+			)
+			return o, s
+		},
+		decGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(2, &o.u32),
+				tlv.MakePrimitiveRecord(3, &o.u64),
+				tlv.MakePrimitiveRecord(4, &o.buf),
+			)
+			return o, s
+		},
+		expObj: &testObj{
+			u32: 3,
+			buf: bytes.Repeat([]byte{0x05}, 3),
+		},
+	},
+	{
+		name: "partial decode",
+		encGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{
+				u16: 2,
+				u32: 5,
+				u64: 7,
+				buf: bytes.Repeat([]byte{0x05}, 3),
+			}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(3, &o.u32),
+				tlv.MakePrimitiveRecord(5, &o.u64),
+				tlv.MakePrimitiveRecord(7, &o.buf),
+			)
+			return o, s
+		},
+		decGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(5, &o.u64),
+			)
+			return o, s
+		},
+		expObj: &testObj{
+			u16: 2,
+			u64: 7,
+		},
+	},
+	{
```

Comment:
Another test case may be that there is no overlap at all between encode and decode types